File: programmers/프로그래머스_베스트앨범.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>
#include <iostream>
using namespace std;

vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;
    int value;
    map<string,int> m;
    vector <pair<int,string>> v;
    map<string,vector<pair<int,int>> > temp;
    
    for(int i=0;i<genres.size();i++){
        if(m.find(genres[i])==m.end()) m[genres[i]]=plays[i];
        else m[genres[i]]+=plays[i];
        temp[genres[i]].push_back({plays[i],-i});
    }
    for(auto iter=m.begin();iter!=m.end();++iter) v.push_back({iter->second,iter->first});
    sort(v.begin(),v.end(),greater<pair<int,string>>());
    for(auto iter=temp.begin();iter!=temp.end();++iter) sort(iter->second.begin(),iter->second.end(),greater<pair<int,int>>());
    for(int i=0;i<v.size();i++){
        for(int j=0;j<min(2,(int)temp[v[i].second].size());j++) answer.push_back(-temp[v[i].second][j].second);
    }
    return answer;
}

File: programmers/프로그래머스_베스트앨범.py
from collections import defaultdict
def solution(genres, plays):
    answer = []
    totalPlay=defaultdict(int)
    totalPlayedSong=defaultdict(list)
    for i,play in enumerate(plays):
        totalPlay[genres[i]]+=play
        totalPlayedSong[genres[i]].append((play,i))
    totalPlay=sorted([(value,key) for key,value in totalPlay.items()],reverse=True)
    for _,genres in totalPlay:
        temp=sorted(totalPlayedSong[genres],key=lambda x:(-x[0],x[1]))
        answer.extend([j for i,j in temp[:2] ]) 
    return answer
